diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9e786154a2b4..b8dcf1b94eb7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -157,8 +157,11 @@ set(MLIR_INSTALL_AGGREGATE_OBJECTS 1 CACHE BOOL
 
 set(MLIR_BUILD_MLIR_C_DYLIB 0 CACHE BOOL "Builds libMLIR-C shared library.")
 
-set(MLIR_LINK_MLIR_DYLIB ${LLVM_LINK_LLVM_DYLIB} CACHE BOOL
-    "Link tools against libMLIR.so")
+set(MLIR_BUILD_MLIR_DYLIB 0 CACHE BOOL "Builds the libMLIR shared library")
+set(MLIR_LINK_MLIR_DYLIB ${LLVM_LINK_LLVM_DYLIB} CACHE BOOL "Link tools against libMLIR.so")
+if (MLIR_LINK_MLIR_DYLIB)
+  set(MLIR_BUILD_MLIR_DYLIB 1)
+endif()
 
 configure_file(
   ${MLIR_MAIN_INCLUDE_DIR}/mlir/Config/mlir-config.h.cmake
@@ -217,6 +220,15 @@ set(MLIR_PDLL_TABLEGEN_TARGET "${MLIR_PDLL_TABLEGEN_TARGET}" CACHE INTERNAL "")
 set(MLIR_SRC_SHARDER_TABLEGEN_EXE "${MLIR_SRC_SHARDER_TABLEGEN_EXE}" CACHE INTERNAL "")
 set(MLIR_SRC_SHARDER_TABLEGEN_TARGET "${MLIR_SRC_SHARDER_TABLEGEN_TARGET}" CACHE INTERNAL "")
 
+# Add MLIR dylib target here, as calls to mlir_target_link_libraries (used by
+# individual libs below) assume that such a target may exist,
+# but we cannot define the actual dylib until after all individual libs
+# are defined.
+if (MLIR_LINK_MLIR_DYLIB)
+  add_library(MLIRDylib INTERFACE)
+  add_mlir_library_install(MLIRDylib)
+endif()
+
 add_subdirectory(include/mlir)
 add_subdirectory(lib)
 # C API needs all dialects for registration, but should be built before tests.
diff --git a/cmake/modules/AddMLIR.cmake b/cmake/modules/AddMLIR.cmake
index 2fc84a5b2879..f614a9eeb34b 100644
--- a/cmake/modules/AddMLIR.cmake
+++ b/cmake/modules/AddMLIR.cmake
@@ -196,6 +196,11 @@ function(add_mlir_interface interface)
   add_dependencies(mlir-generic-headers MLIR${interface}IncGen)
 endfunction()
 
+# Add a tablegen target that generates headers in the include directory
+macro(add_mlir_tablegen_target target)
+  add_public_tablegen_target(${target})
+  add_dependencies(mlir-headers ${target})
+endmacro()
 
 # Generate Documentation
 function(add_mlir_doc doc_filename output_file output_directory command)
@@ -347,7 +352,7 @@ function(add_mlir_library name)
     # Yes, because the target "obj.${name}" is referenced.
     set(NEEDS_OBJECT_LIB ON)
   endif ()
-  if(LLVM_BUILD_LLVM_DYLIB AND NOT ARG_EXCLUDE_FROM_LIBMLIR AND NOT XCODE)
+  if(MLIR_BUILD_MLIR_DYLIB AND NOT ARG_EXCLUDE_FROM_LIBMLIR AND NOT XCODE)
     # Yes, because in addition to the shared library, the object files are
     # needed for linking into libMLIR.so (see mlir/tools/mlir-shlib/CMakeLists.txt).
     # For XCode, -force_load is used instead.
@@ -734,7 +739,7 @@ function(mlir_target_link_libraries target type)
   endif()
 
   if (MLIR_LINK_MLIR_DYLIB)
-    target_link_libraries(${target} ${type} MLIR)
+    target_link_libraries(${target} ${type} MLIRDylib)
   else()
     target_link_libraries(${target} ${type} ${ARGN})
   endif()
diff --git a/include/mlir/Conversion/CMakeLists.txt b/include/mlir/Conversion/CMakeLists.txt
index 9f76ab659215..0a19b88e29dc 100644
--- a/include/mlir/Conversion/CMakeLists.txt
+++ b/include/mlir/Conversion/CMakeLists.txt
@@ -4,6 +4,7 @@ mlir_tablegen(Passes.h.inc -gen-pass-decls -name Conversion)
 mlir_tablegen(Passes.capi.h.inc -gen-pass-capi-header --prefix Conversion)
 mlir_tablegen(Passes.capi.cpp.inc -gen-pass-capi-impl --prefix Conversion)
 add_public_tablegen_target(MLIRConversionPassIncGen)
+add_dependencies(mlir-generic-headers MLIRConversionPassIncGen)
 
 add_mlir_doc(Passes ConversionPasses ./ -gen-pass-doc)
 
diff --git a/include/mlir/Dialect/AMDGPU/IR/CMakeLists.txt b/include/mlir/Dialect/AMDGPU/IR/CMakeLists.txt
index ed074c205a55..e19e72e8e603 100644
--- a/include/mlir/Dialect/AMDGPU/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/AMDGPU/IR/CMakeLists.txt
@@ -4,9 +4,9 @@ add_mlir_doc(AMDGPU AMDGPU Dialects/ -gen-dialect-doc)
 set(LLVM_TARGET_DEFINITIONS AMDGPU.td)
 mlir_tablegen(AMDGPUEnums.h.inc -gen-enum-decls)
 mlir_tablegen(AMDGPUEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRAMDGPUEnumsGen)
+add_mlir_tablegen_target(MLIRAMDGPUEnumsGen)
 
 set(LLVM_TARGET_DEFINITIONS AMDGPU.td)
 mlir_tablegen(AMDGPUAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=amdgpu)
 mlir_tablegen(AMDGPUAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=amdgpu)
-add_public_tablegen_target(MLIRAMDGPUAttributesIncGen)
+add_mlir_tablegen_target(MLIRAMDGPUAttributesIncGen)
diff --git a/include/mlir/Dialect/AMX/CMakeLists.txt b/include/mlir/Dialect/AMX/CMakeLists.txt
index f3f1aff5a636..6ff6cc527af9 100644
--- a/include/mlir/Dialect/AMX/CMakeLists.txt
+++ b/include/mlir/Dialect/AMX/CMakeLists.txt
@@ -3,4 +3,4 @@ add_mlir_doc(AMX AMX Dialects/ -gen-dialect-doc -dialect=amx)
 
 set(LLVM_TARGET_DEFINITIONS AMX.td)
 mlir_tablegen(AMXConversions.inc -gen-llvmir-conversions)
-add_public_tablegen_target(MLIRAMXConversionsIncGen)
+add_mlir_tablegen_target(MLIRAMXConversionsIncGen)
diff --git a/include/mlir/Dialect/Affine/CMakeLists.txt b/include/mlir/Dialect/Affine/CMakeLists.txt
index fe1b372f6e03..2327f4ad65ef 100644
--- a/include/mlir/Dialect/Affine/CMakeLists.txt
+++ b/include/mlir/Dialect/Affine/CMakeLists.txt
@@ -3,6 +3,6 @@ add_subdirectory(TransformOps)
 
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Affine)
-add_public_tablegen_target(MLIRAffinePassIncGen)
+add_mlir_tablegen_target(MLIRAffinePassIncGen)
 
 add_mlir_doc(Passes AffinePasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Affine/TransformOps/CMakeLists.txt b/include/mlir/Dialect/Affine/TransformOps/CMakeLists.txt
index c743f5c0766b..466a4b74dbb5 100644
--- a/include/mlir/Dialect/Affine/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/Affine/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS AffineTransformOps.td)
 mlir_tablegen(AffineTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(AffineTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRAffineTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRAffineTransformOpsIncGen)
 
 add_mlir_doc(AffineTransformOps AffineLoopTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/ArmNeon/CMakeLists.txt b/include/mlir/Dialect/ArmNeon/CMakeLists.txt
index 1c679bcd049b..91581c2ded04 100644
--- a/include/mlir/Dialect/ArmNeon/CMakeLists.txt
+++ b/include/mlir/Dialect/ArmNeon/CMakeLists.txt
@@ -3,4 +3,4 @@ add_mlir_doc(ArmNeon ArmNeon Dialects/ -gen-dialect-doc -dialect=arm_neon)
 
 set(LLVM_TARGET_DEFINITIONS ArmNeon.td)
 mlir_tablegen(ArmNeonConversions.inc -gen-llvmir-conversions)
-add_public_tablegen_target(MLIRArmNeonConversionsIncGen)
+add_mlir_tablegen_target(MLIRArmNeonConversionsIncGen)
diff --git a/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt b/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt
index 9801d8b099e3..1e5c37a20417 100644
--- a/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt
@@ -8,25 +8,25 @@ mlir_tablegen(ArmSMEEnums.h.inc -gen-enum-decls)
 mlir_tablegen(ArmSMEEnums.cpp.inc -gen-enum-defs)
 mlir_tablegen(ArmSMEAttrDefs.h.inc -gen-attrdef-decls -attrdefs-dialect=arm_sme)
 mlir_tablegen(ArmSMEAttrDefs.cpp.inc -gen-attrdef-defs -attrdefs-dialect=arm_sme)
-add_public_tablegen_target(MLIRArmSMEOpsIncGen)
+add_mlir_tablegen_target(MLIRArmSMEOpsIncGen)
 
 # Generate LLVM IR Conversions
 set(LLVM_TARGET_DEFINITIONS ArmSMEOps.td)
 mlir_tablegen(ArmSMEOpsConversions.inc -gen-llvmir-conversions)
-add_public_tablegen_target(MLIRArmSMEConversionsIncGen)
+add_mlir_tablegen_target(MLIRArmSMEConversionsIncGen)
 
 # Generate op interface declarations and definitions
 set(LLVM_TARGET_DEFINITIONS ArmSMEOps.td)
 mlir_tablegen(ArmSMEOpInterfaces.h.inc -gen-op-interface-decls)
 mlir_tablegen(ArmSMEOpInterfaces.cpp.inc -gen-op-interface-defs)
-add_public_tablegen_target(MLIRArmSMEOpInterfaces)
+add_mlir_tablegen_target(MLIRArmSMEOpInterfaces)
 
 # Generate declarations and definitions of ArmSME intrinsic Ops
 set(LLVM_TARGET_DEFINITIONS ArmSMEIntrinsicOps.td)
 mlir_tablegen(ArmSMEIntrinsicOps.h.inc -gen-op-decls)
 mlir_tablegen(ArmSMEIntrinsicOps.cpp.inc -gen-op-defs)
 mlir_tablegen(ArmSMEIntrinsicConversions.inc -gen-llvmir-conversions)
-add_public_tablegen_target(MLIRArmSMEIntrinsicOpsIncGen)
+add_mlir_tablegen_target(MLIRArmSMEIntrinsicOpsIncGen)
 
 # Generate the docs
 add_mlir_doc(ArmSMEOps ArmSMEOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/ArmSVE/IR/CMakeLists.txt b/include/mlir/Dialect/ArmSVE/IR/CMakeLists.txt
index 06595b7088a1..f8d06951e727 100644
--- a/include/mlir/Dialect/ArmSVE/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/ArmSVE/IR/CMakeLists.txt
@@ -3,4 +3,4 @@ add_mlir_doc(ArmSVE ArmSVE Dialects/ -gen-dialect-doc -dialect=arm_sve)
 
 set(LLVM_TARGET_DEFINITIONS ArmSVE.td)
 mlir_tablegen(ArmSVEConversions.inc -gen-llvmir-conversions)
-add_public_tablegen_target(MLIRArmSVEConversionsIncGen)
+add_mlir_tablegen_target(MLIRArmSVEConversionsIncGen)
diff --git a/include/mlir/Dialect/Async/CMakeLists.txt b/include/mlir/Dialect/Async/CMakeLists.txt
index cabd5d3087cf..f2642d646283 100644
--- a/include/mlir/Dialect/Async/CMakeLists.txt
+++ b/include/mlir/Dialect/Async/CMakeLists.txt
@@ -4,6 +4,6 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Async)
 mlir_tablegen(Passes.capi.h.inc -gen-pass-capi-header --prefix Async)
 mlir_tablegen(Passes.capi.cpp.inc -gen-pass-capi-impl --prefix Async)
-add_public_tablegen_target(MLIRAsyncPassIncGen)
+add_mlir_tablegen_target(MLIRAsyncPassIncGen)
 
 add_mlir_doc(Passes AsyncPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Bufferization/IR/CMakeLists.txt b/include/mlir/Dialect/Bufferization/IR/CMakeLists.txt
index 13a5bc370a4f..6a71f4355bf2 100644
--- a/include/mlir/Dialect/Bufferization/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Bufferization/IR/CMakeLists.txt
@@ -8,5 +8,4 @@ add_mlir_interface(BufferViewFlowOpInterface)
 set(LLVM_TARGET_DEFINITIONS BufferizationEnums.td)
 mlir_tablegen(BufferizationEnums.h.inc -gen-enum-decls)
 mlir_tablegen(BufferizationEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRBufferizationEnumsIncGen)
-add_dependencies(mlir-headers MLIRBufferizationEnumsIncGen)
+add_mlir_tablegen_target(MLIRBufferizationEnumsIncGen)
diff --git a/include/mlir/Dialect/Bufferization/TransformOps/CMakeLists.txt b/include/mlir/Dialect/Bufferization/TransformOps/CMakeLists.txt
index 95276e3e5e0d..684c4295cd4f 100644
--- a/include/mlir/Dialect/Bufferization/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/Bufferization/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS BufferizationTransformOps.td)
 mlir_tablegen(BufferizationTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(BufferizationTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRBufferizationTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRBufferizationTransformOpsIncGen)
 
 add_mlir_doc(BufferizationTransformOps BufferizationTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Bufferization/Transforms/CMakeLists.txt b/include/mlir/Dialect/Bufferization/Transforms/CMakeLists.txt
index dcae4b8f26b8..bd6ea9a1cb29 100644
--- a/include/mlir/Dialect/Bufferization/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Bufferization/Transforms/CMakeLists.txt
@@ -1,6 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Bufferization)
-add_public_tablegen_target(MLIRBufferizationPassIncGen)
-add_dependencies(mlir-headers MLIRBufferizationPassIncGen)
+add_mlir_tablegen_target(MLIRBufferizationPassIncGen)
 
 add_mlir_doc(Passes BufferizationPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Complex/IR/CMakeLists.txt b/include/mlir/Dialect/Complex/IR/CMakeLists.txt
index f41888d01a2f..abffb2ac3fcf 100644
--- a/include/mlir/Dialect/Complex/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Complex/IR/CMakeLists.txt
@@ -4,4 +4,4 @@ add_mlir_doc(ComplexOps ComplexOps Dialects/ -gen-dialect-doc  -dialect=complex)
 set(LLVM_TARGET_DEFINITIONS ComplexAttributes.td)
 mlir_tablegen(ComplexAttributes.h.inc -gen-attrdef-decls)
 mlir_tablegen(ComplexAttributes.cpp.inc -gen-attrdef-defs)
-add_public_tablegen_target(MLIRComplexAttributesIncGen)
+add_mlir_tablegen_target(MLIRComplexAttributesIncGen)
diff --git a/include/mlir/Dialect/DLTI/CMakeLists.txt b/include/mlir/Dialect/DLTI/CMakeLists.txt
index 4f8382e8e6e6..74b1c541edae 100644
--- a/include/mlir/Dialect/DLTI/CMakeLists.txt
+++ b/include/mlir/Dialect/DLTI/CMakeLists.txt
@@ -6,5 +6,4 @@ add_mlir_doc(DLTIAttrs DLTIDialect Dialects/ -gen-dialect-doc)
 set(LLVM_TARGET_DEFINITIONS DLTIAttrs.td)
 mlir_tablegen(DLTIAttrs.h.inc -gen-attrdef-decls -attrdefs-dialect=dlti)
 mlir_tablegen(DLTIAttrs.cpp.inc -gen-attrdef-defs -attrdefs-dialect=dlti)
-add_public_tablegen_target(MLIRDLTIAttrsIncGen)
-add_dependencies(mlir-headers MLIRDLTIAttrsIncGen)
+add_mlir_tablegen_target(MLIRDLTIAttrsIncGen)
diff --git a/include/mlir/Dialect/DLTI/TransformOps/CMakeLists.txt b/include/mlir/Dialect/DLTI/TransformOps/CMakeLists.txt
index 1188d1a9b24d..57faeca06794 100644
--- a/include/mlir/Dialect/DLTI/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/DLTI/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS DLTITransformOps.td)
 mlir_tablegen(DLTITransformOps.h.inc -gen-op-decls)
 mlir_tablegen(DLTITransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRDLTITransformOpsIncGen)
+add_mlir_tablegen_target(MLIRDLTITransformOpsIncGen)
 
 add_mlir_doc(DLTITransformOps DLTITransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/EmitC/IR/CMakeLists.txt b/include/mlir/Dialect/EmitC/IR/CMakeLists.txt
index ac8c651cdced..cb25f555d151 100644
--- a/include/mlir/Dialect/EmitC/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/EmitC/IR/CMakeLists.txt
@@ -6,4 +6,4 @@ mlir_tablegen(EmitCEnums.h.inc -gen-enum-decls)
 mlir_tablegen(EmitCEnums.cpp.inc -gen-enum-defs)
 mlir_tablegen(EmitCAttributes.h.inc -gen-attrdef-decls)
 mlir_tablegen(EmitCAttributes.cpp.inc -gen-attrdef-defs)
-add_public_tablegen_target(MLIREmitCAttributesIncGen)
+add_mlir_tablegen_target(MLIREmitCAttributesIncGen)
diff --git a/include/mlir/Dialect/EmitC/Transforms/CMakeLists.txt b/include/mlir/Dialect/EmitC/Transforms/CMakeLists.txt
index 0b507d75fa07..d112be898c35 100644
--- a/include/mlir/Dialect/EmitC/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/EmitC/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name EmitC)
-add_public_tablegen_target(MLIREmitCTransformsIncGen)
+add_mlir_tablegen_target(MLIREmitCTransformsIncGen)
 
 add_mlir_doc(Passes EmitCPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/GPU/IR/CMakeLists.txt b/include/mlir/Dialect/GPU/IR/CMakeLists.txt
index 3c95b4f8fb0d..eb873f12315c 100644
--- a/include/mlir/Dialect/GPU/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/GPU/IR/CMakeLists.txt
@@ -4,29 +4,29 @@ add_mlir_doc(GPUOps GPUOps Dialects/ -gen-op-doc)
 set(LLVM_TARGET_DEFINITIONS GPUBase.td)
 mlir_tablegen(GPUOpInterfaces.h.inc -gen-op-interface-decls)
 mlir_tablegen(GPUOpInterfaces.cpp.inc -gen-op-interface-defs)
-add_public_tablegen_target(MLIRGPUOpInterfacesIncGen)
+add_mlir_tablegen_target(MLIRGPUOpInterfacesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS ParallelLoopMapperAttr.td)
 mlir_tablegen(ParallelLoopMapperEnums.h.inc -gen-enum-decls)
 mlir_tablegen(ParallelLoopMapperEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRParallelLoopMapperEnumsGen)
+add_mlir_tablegen_target(MLIRParallelLoopMapperEnumsGen)
 
 set(LLVM_TARGET_DEFINITIONS GPUOps.td)
 mlir_tablegen(GPUOpsEnums.h.inc -gen-enum-decls)
 mlir_tablegen(GPUOpsEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRGPUOpsEnumsGen)
+add_mlir_tablegen_target(MLIRGPUOpsEnumsGen)
 
 set(LLVM_TARGET_DEFINITIONS CompilationAttrInterfaces.td)
 mlir_tablegen(CompilationAttrInterfaces.h.inc -gen-attr-interface-decls)
 mlir_tablegen(CompilationAttrInterfaces.cpp.inc -gen-attr-interface-defs)
-add_public_tablegen_target(MLIRGPUCompilationAttrInterfacesIncGen)
+add_mlir_tablegen_target(MLIRGPUCompilationAttrInterfacesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS GPUDeviceMappingAttr.td)
 mlir_tablegen(GPUDeviceMapperEnums.h.inc -gen-enum-decls)
 mlir_tablegen(GPUDeviceMapperEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRGPUDeviceMapperEnumsGen)
+add_mlir_tablegen_target(MLIRGPUDeviceMapperEnumsGen)
 
 set(LLVM_TARGET_DEFINITIONS GPUOps.td)
 mlir_tablegen(GPUOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=gpu)
 mlir_tablegen(GPUOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=gpu)
-add_public_tablegen_target(MLIRGPUOpsAttributesIncGen)
+add_mlir_tablegen_target(MLIRGPUOpsAttributesIncGen)
diff --git a/include/mlir/Dialect/GPU/TransformOps/CMakeLists.txt b/include/mlir/Dialect/GPU/TransformOps/CMakeLists.txt
index c99f3dfa5683..a81a4d60bf73 100644
--- a/include/mlir/Dialect/GPU/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/GPU/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS GPUTransformOps.td)
 mlir_tablegen(GPUTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(GPUTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRGPUTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRGPUTransformOpsIncGen)
 
 add_mlir_doc(GPUTransformOps GPUTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/GPU/Transforms/CMakeLists.txt b/include/mlir/Dialect/GPU/Transforms/CMakeLists.txt
index 60daed4ee97e..ca4fc5ae3d4b 100644
--- a/include/mlir/Dialect/GPU/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/GPU/Transforms/CMakeLists.txt
@@ -2,6 +2,6 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name GPU)
 mlir_tablegen(Passes.capi.h.inc -gen-pass-capi-header --prefix GPU)
 mlir_tablegen(Passes.capi.cpp.inc -gen-pass-capi-impl --prefix GPU)
-add_public_tablegen_target(MLIRGPUPassIncGen)
+add_mlir_tablegen_target(MLIRGPUPassIncGen)
 
 add_mlir_doc(Passes GPUPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/LLVMIR/CMakeLists.txt b/include/mlir/Dialect/LLVMIR/CMakeLists.txt
index 759de745440c..170fb036fb52 100644
--- a/include/mlir/Dialect/LLVMIR/CMakeLists.txt
+++ b/include/mlir/Dialect/LLVMIR/CMakeLists.txt
@@ -11,17 +11,17 @@ mlir_tablegen(LLVMOpsAttrDefs.h.inc -gen-attrdef-decls
               -attrdefs-dialect=llvm)
 mlir_tablegen(LLVMOpsAttrDefs.cpp.inc -gen-attrdef-defs
               -attrdefs-dialect=llvm)
-add_public_tablegen_target(MLIRLLVMOpsIncGen)
+add_mlir_tablegen_target(MLIRLLVMOpsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LLVMTypes.td)
 mlir_tablegen(LLVMTypes.h.inc -gen-typedef-decls -typedefs-dialect=llvm)
 mlir_tablegen(LLVMTypes.cpp.inc -gen-typedef-defs -typedefs-dialect=llvm)
-add_public_tablegen_target(MLIRLLVMTypesIncGen)
+add_mlir_tablegen_target(MLIRLLVMTypesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LLVMIntrinsicOps.td)
 mlir_tablegen(LLVMIntrinsicOps.h.inc -gen-op-decls)
 mlir_tablegen(LLVMIntrinsicOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRLLVMIntrinsicOpsIncGen)
+add_mlir_tablegen_target(MLIRLLVMIntrinsicOpsIncGen)
 
 add_mlir_doc(LLVMOps LLVMOps Dialects/ -gen-op-doc)
 add_mlir_doc(LLVMIntrinsicOps LLVMIntrinsicOps Dialects/ -gen-op-doc)
@@ -33,25 +33,25 @@ mlir_tablegen(LLVMAttrInterfaces.h.inc -gen-attr-interface-decls)
 mlir_tablegen(LLVMAttrInterfaces.cpp.inc -gen-attr-interface-defs)
 mlir_tablegen(LLVMTypeInterfaces.h.inc -gen-type-interface-decls)
 mlir_tablegen(LLVMTypeInterfaces.cpp.inc -gen-type-interface-defs)
-add_public_tablegen_target(MLIRLLVMInterfacesIncGen)
+add_mlir_tablegen_target(MLIRLLVMInterfacesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LLVMOps.td)
 mlir_tablegen(LLVMConversions.inc -gen-llvmir-conversions)
 mlir_tablegen(LLVMConversionEnumsToLLVM.inc -gen-enum-to-llvmir-conversions)
 mlir_tablegen(LLVMConversionEnumsFromLLVM.inc -gen-enum-from-llvmir-conversions)
 mlir_tablegen(LLVMOpFromLLVMIRConversions.inc -gen-op-from-llvmir-conversions)
-add_public_tablegen_target(MLIRLLVMConversionsIncGen)
+add_mlir_tablegen_target(MLIRLLVMConversionsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LLVMIntrinsicOps.td)
 mlir_tablegen(LLVMIntrinsicConversions.inc -gen-llvmir-conversions)
 mlir_tablegen(LLVMIntrinsicFromLLVMIRConversions.inc -gen-intr-from-llvmir-conversions)
 mlir_tablegen(LLVMConvertibleLLVMIRIntrinsics.inc -gen-convertible-llvmir-intrinsics)
-add_public_tablegen_target(MLIRLLVMIntrinsicConversionsIncGen)
+add_mlir_tablegen_target(MLIRLLVMIntrinsicConversionsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BasicPtxBuilderInterface.td)
 mlir_tablegen(BasicPtxBuilderInterface.h.inc -gen-op-interface-decls)
 mlir_tablegen(BasicPtxBuilderInterface.cpp.inc -gen-op-interface-defs)
-add_public_tablegen_target(MLIRBasicPtxBuilderInterfaceIncGen)
+add_mlir_tablegen_target(MLIRBasicPtxBuilderInterfaceIncGen)
 add_dependencies(mlir-headers MLIRBasicPtxBuilderInterfaceIncGen)
 
 add_mlir_dialect(NVVMOps nvvm)
@@ -64,7 +64,7 @@ mlir_tablegen(NVVMOpsEnums.h.inc -gen-enum-decls)
 mlir_tablegen(NVVMOpsEnums.cpp.inc -gen-enum-defs)
 mlir_tablegen(NVVMOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=nvvm)
 mlir_tablegen(NVVMOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=nvvm)
-add_public_tablegen_target(MLIRNVVMConversionsIncGen)
+add_mlir_tablegen_target(MLIRNVVMConversionsIncGen)
 
 add_mlir_dialect(ROCDLOps rocdl)
 add_mlir_doc(ROCDLOps ROCDLDialect Dialects/ -gen-dialect-doc -dialect=rocdl)
@@ -72,7 +72,7 @@ set(LLVM_TARGET_DEFINITIONS ROCDLOps.td)
 mlir_tablegen(ROCDLConversions.inc -gen-llvmir-conversions)
 mlir_tablegen(ROCDLOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=rocdl)
 mlir_tablegen(ROCDLOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=rocdl)
-add_public_tablegen_target(MLIRROCDLConversionsIncGen)
+add_mlir_tablegen_target(MLIRROCDLConversionsIncGen)
 
 add_mlir_dialect(VCIXOps vcix)
 add_mlir_doc(VCIXOps VCIXDialect Dialects/ -gen-dialect-doc -dialect=vcix)
@@ -80,4 +80,4 @@ set(LLVM_TARGET_DEFINITIONS VCIXOps.td)
 mlir_tablegen(VCIXConversions.inc -gen-llvmir-conversions)
 mlir_tablegen(VCIXOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=vcix)
 mlir_tablegen(VCIXOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=vcix)
-add_public_tablegen_target(MLIRVCIXConversionsIncGen)
+add_mlir_tablegen_target(MLIRVCIXConversionsIncGen)
diff --git a/include/mlir/Dialect/LLVMIR/Transforms/CMakeLists.txt b/include/mlir/Dialect/LLVMIR/Transforms/CMakeLists.txt
index becae6a95393..a82894d1075f 100644
--- a/include/mlir/Dialect/LLVMIR/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/LLVMIR/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name LLVM)
-add_public_tablegen_target(MLIRLLVMPassIncGen)
+add_mlir_tablegen_target(MLIRLLVMPassIncGen)
 
 add_mlir_doc(Passes LLVMPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Linalg/CMakeLists.txt b/include/mlir/Dialect/Linalg/CMakeLists.txt
index f0a486df9842..48f3a24ebf94 100644
--- a/include/mlir/Dialect/Linalg/CMakeLists.txt
+++ b/include/mlir/Dialect/Linalg/CMakeLists.txt
@@ -5,6 +5,6 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Linalg)
 mlir_tablegen(Passes.capi.h.inc -gen-pass-capi-header --prefix Linalg)
 mlir_tablegen(Passes.capi.cpp.inc -gen-pass-capi-impl --prefix Linalg)
-add_public_tablegen_target(MLIRLinalgPassIncGen)
+add_mlir_tablegen_target(MLIRLinalgPassIncGen)
 
 add_mlir_doc(Passes LinalgPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Linalg/IR/CMakeLists.txt b/include/mlir/Dialect/Linalg/IR/CMakeLists.txt
index 71214b4404c5..1bc7cb9d3600 100644
--- a/include/mlir/Dialect/Linalg/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Linalg/IR/CMakeLists.txt
@@ -46,14 +46,12 @@ add_mlir_dialect(LinalgOps linalg)
 set(LLVM_TARGET_DEFINITIONS LinalgEnums.td)
 mlir_tablegen(LinalgOpsEnums.h.inc -gen-enum-decls)
 mlir_tablegen(LinalgOpsEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRLinalgOpsEnumsIncGen)
-add_dependencies(mlir-headers MLIRLinalgOpsEnumsIncGen)
+add_mlir_tablegen_target(MLIRLinalgOpsEnumsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LinalgOps.td)
 mlir_tablegen(LinalgOpsAttrDefs.h.inc -gen-attrdef-decls)
 mlir_tablegen(LinalgOpsAttrDefs.cpp.inc -gen-attrdef-defs)
-add_public_tablegen_target(MLIRLinalgOpsAttributesIncGen)
-add_dependencies(mlir-headers MLIRLinalgOpsAttributesIncGen)
+add_mlir_tablegen_target(MLIRLinalgOpsAttributesIncGen)
 
 add_mlir_doc(LinalgDoc LinalgOps Dialects/ -gen-op-doc)
 add_dependencies(LinalgOpsDocGen LinalgOdsGen)
@@ -61,12 +59,10 @@ add_dependencies(LinalgOpsDocGen LinalgOdsGen)
 set(LLVM_TARGET_DEFINITIONS LinalgStructuredOps.td)
 mlir_tablegen(LinalgStructuredOps.h.inc -gen-op-decls)
 mlir_tablegen(LinalgStructuredOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRLinalgStructuredOpsIncGen)
+add_mlir_tablegen_target(MLIRLinalgStructuredOpsIncGen)
 add_dependencies(MLIRLinalgStructuredOpsIncGen LinalgOdsGen)
-add_dependencies(mlir-headers MLIRLinalgStructuredOpsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LinalgInterfaces.td)
 mlir_tablegen(LinalgInterfaces.h.inc -gen-op-interface-decls)
 mlir_tablegen(LinalgInterfaces.cpp.inc -gen-op-interface-defs)
-add_public_tablegen_target(MLIRLinalgInterfacesIncGen)
-add_dependencies(mlir-headers MLIRLinalgInterfacesIncGen)
+add_mlir_tablegen_target(MLIRLinalgInterfacesIncGen)
diff --git a/include/mlir/Dialect/Linalg/TransformOps/CMakeLists.txt b/include/mlir/Dialect/Linalg/TransformOps/CMakeLists.txt
index 4f6b251bf696..6b0cdf2b276a 100644
--- a/include/mlir/Dialect/Linalg/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/Linalg/TransformOps/CMakeLists.txt
@@ -1,17 +1,17 @@
 set(LLVM_TARGET_DEFINITIONS LinalgMatchOps.td)
 mlir_tablegen(LinalgMatchOps.h.inc -gen-op-decls)
 mlir_tablegen(LinalgMatchOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRLinalgMatchOpsIncGen)
+add_mlir_tablegen_target(MLIRLinalgMatchOpsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LinalgTransformOps.td)
 mlir_tablegen(LinalgTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(LinalgTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRLinalgTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRLinalgTransformOpsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS LinalgTransformEnums.td)
 mlir_tablegen(LinalgTransformOpsEnums.h.inc -gen-enum-decls)
 mlir_tablegen(LinalgTransformOpsEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRLinalgTransformEnumsIncGen)
+add_mlir_tablegen_target(MLIRLinalgTransformEnumsIncGen)
 
 add_mlir_doc(LinalgMatchOps LinalgStructuredMatchOps Dialects/ -gen-op-doc)
 add_mlir_doc(LinalgTransformOps LinalgStructuredTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/MLProgram/Transforms/CMakeLists.txt b/include/mlir/Dialect/MLProgram/Transforms/CMakeLists.txt
index c5c11f17a9fa..5118dd952f91 100644
--- a/include/mlir/Dialect/MLProgram/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/MLProgram/Transforms/CMakeLists.txt
@@ -1,6 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name MLProgram)
-add_public_tablegen_target(MLIRMLProgramPassIncGen)
-add_dependencies(mlir-headers MLIRMLProgramPassIncGen)
+add_mlir_tablegen_target(MLIRMLProgramPassIncGen)
 
 add_mlir_doc(Passes MLProgramPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/MPI/IR/CMakeLists.txt b/include/mlir/Dialect/MPI/IR/CMakeLists.txt
index dfec2ea486cb..598c485a3499 100644
--- a/include/mlir/Dialect/MPI/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/MPI/IR/CMakeLists.txt
@@ -5,13 +5,13 @@ add_mlir_doc(MPIOps MPI Dialects/ -gen-dialect-doc)
 set(LLVM_TARGET_DEFINITIONS MPIOps.td)
 mlir_tablegen(MPIOps.h.inc -gen-op-decls)
 mlir_tablegen(MPIOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRMPIOpsIncGen)
+add_mlir_tablegen_target(MLIRMPIOpsIncGen)
 
 # Add MPI types
 set(LLVM_TARGET_DEFINITIONS MPITypes.td)
 mlir_tablegen(MPITypesGen.h.inc -gen-typedef-decls)
 mlir_tablegen(MPITypesGen.cpp.inc -gen-typedef-defs)
-add_public_tablegen_target(MLIRMPITypesIncGen)
+add_mlir_tablegen_target(MLIRMPITypesIncGen)
 
 # Add MPI attributes
 set(LLVM_TARGET_DEFINITIONS MPI.td)
@@ -19,4 +19,4 @@ mlir_tablegen(MPIEnums.h.inc -gen-enum-decls)
 mlir_tablegen(MPIEnums.cpp.inc -gen-enum-defs)
 mlir_tablegen(MPIAttrDefs.h.inc -gen-attrdef-decls)
 mlir_tablegen(MPIAttrDefs.cpp.inc -gen-attrdef-defs)
-add_public_tablegen_target(MLIRMPIAttrsIncGen)
+add_mlir_tablegen_target(MLIRMPIAttrsIncGen)
diff --git a/include/mlir/Dialect/Math/Transforms/CMakeLists.txt b/include/mlir/Dialect/Math/Transforms/CMakeLists.txt
index a37f069da46b..e2ca58faf879 100644
--- a/include/mlir/Dialect/Math/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Math/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Math)
-add_public_tablegen_target(MLIRMathTransformsIncGen)
+add_mlir_tablegen_target(MLIRMathTransformsIncGen)
 
 add_mlir_doc(Passes MathPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/MemRef/TransformOps/CMakeLists.txt b/include/mlir/Dialect/MemRef/TransformOps/CMakeLists.txt
index 8dbe98802359..bdf472d4f51e 100644
--- a/include/mlir/Dialect/MemRef/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/MemRef/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS MemRefTransformOps.td)
 mlir_tablegen(MemRefTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(MemRefTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRMemRefTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRMemRefTransformOpsIncGen)
 
 add_mlir_doc(MemRefTransformOps MemRefTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/MemRef/Transforms/CMakeLists.txt b/include/mlir/Dialect/MemRef/Transforms/CMakeLists.txt
index 6f868f7be1b2..af57472ea1ae 100644
--- a/include/mlir/Dialect/MemRef/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/MemRef/Transforms/CMakeLists.txt
@@ -1,6 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name MemRef)
-add_public_tablegen_target(MLIRMemRefPassIncGen)
-add_dependencies(mlir-headers MLIRMemRefPassIncGen)
+add_mlir_tablegen_target(MLIRMemRefPassIncGen)
 
 add_mlir_doc(Passes MemRefPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Mesh/IR/CMakeLists.txt b/include/mlir/Dialect/Mesh/IR/CMakeLists.txt
index f26c6285efd8..e6639afcc35f 100644
--- a/include/mlir/Dialect/Mesh/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Mesh/IR/CMakeLists.txt
@@ -21,5 +21,4 @@ set(LLVM_TARGET_DEFINITIONS MeshOps.td)
 mlir_tablegen(MeshOps.h.inc -gen-op-decls)
 mlir_tablegen(MeshOps.cpp.inc -gen-op-defs)
 
-add_public_tablegen_target(MLIRMeshIncGen)
-add_dependencies(mlir-headers MLIRMeshIncGen)
+add_mlir_tablegen_target(MLIRMeshIncGen)
diff --git a/include/mlir/Dialect/Mesh/Interfaces/CMakeLists.txt b/include/mlir/Dialect/Mesh/Interfaces/CMakeLists.txt
index b3a44f3b0089..0f82b808781c 100644
--- a/include/mlir/Dialect/Mesh/Interfaces/CMakeLists.txt
+++ b/include/mlir/Dialect/Mesh/Interfaces/CMakeLists.txt
@@ -1,4 +1,4 @@
 set(LLVM_TARGET_DEFINITIONS ShardingInterface.td)
 mlir_tablegen(ShardingInterface.h.inc -gen-op-interface-decls)
 mlir_tablegen(ShardingInterface.cpp.inc -gen-op-interface-defs)
-add_public_tablegen_target(MLIRShardingInterfaceIncGen)
+add_mlir_tablegen_target(MLIRShardingInterfaceIncGen)
diff --git a/include/mlir/Dialect/Mesh/Transforms/CMakeLists.txt b/include/mlir/Dialect/Mesh/Transforms/CMakeLists.txt
index 8d768485103b..d48f1d3d6a62 100644
--- a/include/mlir/Dialect/Mesh/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Mesh/Transforms/CMakeLists.txt
@@ -1,6 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Mesh)
-add_public_tablegen_target(MLIRMeshPassIncGen)
-add_dependencies(mlir-headers MLIRMeshPassIncGen)
+add_mlir_tablegen_target(MLIRMeshPassIncGen)
 
 add_mlir_doc(Passes MeshPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/NVGPU/IR/CMakeLists.txt b/include/mlir/Dialect/NVGPU/IR/CMakeLists.txt
index 13d754ca0631..cedaa39867fd 100644
--- a/include/mlir/Dialect/NVGPU/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/NVGPU/IR/CMakeLists.txt
@@ -4,14 +4,14 @@ add_mlir_doc(NVGPU NVGPU Dialects/ -gen-dialect-doc)
 set(LLVM_TARGET_DEFINITIONS NVGPU.td)
 mlir_tablegen(NVGPUEnums.h.inc -gen-enum-decls)
 mlir_tablegen(NVGPUEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRNVGPUEnumsIncGen)
+add_mlir_tablegen_target(MLIRNVGPUEnumsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS NVGPU.td)
 mlir_tablegen(NVGPUAttrDefs.h.inc -gen-attrdef-decls)
 mlir_tablegen(NVGPUAttrDefs.cpp.inc -gen-attrdef-defs)
-add_public_tablegen_target(MLIRNVGPUAttributesIncGen)
+add_mlir_tablegen_target(MLIRNVGPUAttributesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS NVGPU.td)
 mlir_tablegen(NVGPUAttrTypes.h.inc -gen-typedef-decls)
 mlir_tablegen(NVGPUAttrTypes.cpp.inc -gen-typedef-decls)
-add_public_tablegen_target(MLIRNVGPUTypesIncGen)
+add_mlir_tablegen_target(MLIRNVGPUTypesIncGen)
diff --git a/include/mlir/Dialect/NVGPU/Transforms/CMakeLists.txt b/include/mlir/Dialect/NVGPU/Transforms/CMakeLists.txt
index 706b66e24b4c..d9d83282f0c8 100644
--- a/include/mlir/Dialect/NVGPU/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/NVGPU/Transforms/CMakeLists.txt
@@ -2,6 +2,6 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name NVGPU)
 mlir_tablegen(Passes.capi.h.inc -gen-pass-capi-header --prefix NVGPU)
 mlir_tablegen(Passes.capi.cpp.inc -gen-pass-capi-impl --prefix NVGPU)
-add_public_tablegen_target(MLIRNVGPUPassIncGen)
+add_mlir_tablegen_target(MLIRNVGPUPassIncGen)
 
 add_mlir_doc(Passes NVGPUPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/OpenACC/CMakeLists.txt b/include/mlir/Dialect/OpenACC/CMakeLists.txt
index 66b1e89a9881..5797b973e3cd 100644
--- a/include/mlir/Dialect/OpenACC/CMakeLists.txt
+++ b/include/mlir/Dialect/OpenACC/CMakeLists.txt
@@ -12,19 +12,16 @@ add_dependencies(OpenACCDialectOpsDocGen acc_common_td)
 set(LLVM_TARGET_DEFINITIONS OpenACCOps.td)
 mlir_tablegen(OpenACCOpsEnums.h.inc -gen-enum-decls)
 mlir_tablegen(OpenACCOpsEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIROpenACCEnumsIncGen)
-add_dependencies(mlir-headers MLIROpenACCEnumsIncGen)
+add_mlir_tablegen_target(MLIROpenACCEnumsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS OpenACCOps.td)
 mlir_tablegen(OpenACCOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=acc)
 mlir_tablegen(OpenACCOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=acc)
-add_public_tablegen_target(MLIROpenACCAttributesIncGen)
-add_dependencies(mlir-headers MLIROpenACCAttributesIncGen)
+add_mlir_tablegen_target(MLIROpenACCAttributesIncGen)
 
 add_mlir_interface(OpenACCOpsInterfaces)
 
 set(LLVM_TARGET_DEFINITIONS OpenACCTypeInterfaces.td)
 mlir_tablegen(OpenACCTypeInterfaces.h.inc -gen-type-interface-decls)
 mlir_tablegen(OpenACCTypeInterfaces.cpp.inc -gen-type-interface-defs)
-add_public_tablegen_target(MLIROpenACCTypeInterfacesIncGen)
-add_dependencies(mlir-headers MLIROpenACCTypeInterfacesIncGen)
+add_mlir_tablegen_target(MLIROpenACCTypeInterfacesIncGen)
diff --git a/include/mlir/Dialect/OpenACC/Transforms/CMakeLists.txt b/include/mlir/Dialect/OpenACC/Transforms/CMakeLists.txt
index ddbd5839576f..b89d2797bdb5 100644
--- a/include/mlir/Dialect/OpenACC/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/OpenACC/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name OpenACC)
-add_public_tablegen_target(MLIROpenACCPassIncGen)
+add_mlir_tablegen_target(MLIROpenACCPassIncGen)
 
 add_mlir_doc(Passes OpenACCPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/OpenMP/CMakeLists.txt b/include/mlir/Dialect/OpenMP/CMakeLists.txt
index a65c6b1d3c96..b8eb551062b6 100644
--- a/include/mlir/Dialect/OpenMP/CMakeLists.txt
+++ b/include/mlir/Dialect/OpenMP/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS ${LLVM_MAIN_INCLUDE_DIR}/llvm/Frontend/OpenMP/OMP.td)
 mlir_tablegen(OmpCommon.td --gen-directive-decl --directives-dialect=OpenMP)
-add_public_tablegen_target(omp_common_td)
+add_mlir_tablegen_target(omp_common_td)
 
 set(LLVM_TARGET_DEFINITIONS OpenMPOps.td)
 
diff --git a/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt b/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt
index ecdea158ddef..c99d8d58bfb6 100644
--- a/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt
@@ -4,4 +4,4 @@ add_mlir_doc(Polynomial PolynomialDialect Dialects/ -gen-dialect-doc -dialect=po
 set(LLVM_TARGET_DEFINITIONS PolynomialAttributes.td)
 mlir_tablegen(PolynomialAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=polynomial)
 mlir_tablegen(PolynomialAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=polynomial)
-add_public_tablegen_target(MLIRPolynomialAttributesIncGen)
+add_mlir_tablegen_target(MLIRPolynomialAttributesIncGen)
diff --git a/include/mlir/Dialect/Ptr/IR/CMakeLists.txt b/include/mlir/Dialect/Ptr/IR/CMakeLists.txt
index 82938962fb93..4b78a926c48f 100644
--- a/include/mlir/Dialect/Ptr/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Ptr/IR/CMakeLists.txt
@@ -4,4 +4,4 @@ add_mlir_doc(PtrOps PtrOps Dialects/ -gen-dialect-doc -dialect=ptr)
 set(LLVM_TARGET_DEFINITIONS PtrOps.td)
 mlir_tablegen(PtrOpsAttrs.h.inc -gen-attrdef-decls -attrdefs-dialect=ptr)
 mlir_tablegen(PtrOpsAttrs.cpp.inc -gen-attrdef-defs -attrdefs-dialect=ptr)
-add_public_tablegen_target(MLIRPtrOpsAttributesIncGen)
+add_mlir_tablegen_target(MLIRPtrOpsAttributesIncGen)
diff --git a/include/mlir/Dialect/Quant/IR/CMakeLists.txt b/include/mlir/Dialect/Quant/IR/CMakeLists.txt
index c08f399ee182..cc7572e33239 100644
--- a/include/mlir/Dialect/Quant/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/Quant/IR/CMakeLists.txt
@@ -3,4 +3,4 @@ add_mlir_doc(QuantOps QuantDialect Dialects/ -gen-dialect-doc)
 
 set(LLVM_TARGET_DEFINITIONS QuantDialectBytecode.td)
 mlir_tablegen(QuantDialectBytecode.cpp.inc -gen-bytecode -bytecode-dialect="Quant")
-add_public_tablegen_target(MLIRQuantDialectBytecodeIncGen)
+add_mlir_tablegen_target(MLIRQuantDialectBytecodeIncGen)
diff --git a/include/mlir/Dialect/SCF/TransformOps/CMakeLists.txt b/include/mlir/Dialect/SCF/TransformOps/CMakeLists.txt
index 9095b1f8dbc9..095c2d6bab72 100644
--- a/include/mlir/Dialect/SCF/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/SCF/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS SCFTransformOps.td)
 mlir_tablegen(SCFTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(SCFTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRSCFTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRSCFTransformOpsIncGen)
 
 add_mlir_doc(SCFTransformOps SCFLoopTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/SCF/Transforms/CMakeLists.txt b/include/mlir/Dialect/SCF/Transforms/CMakeLists.txt
index 1192badea78e..09eb3d032625 100644
--- a/include/mlir/Dialect/SCF/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/SCF/Transforms/CMakeLists.txt
@@ -1,7 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name SCF)
-add_public_tablegen_target(MLIRSCFPassIncGen)
-add_dependencies(mlir-headers MLIRSCFPassIncGen)
+add_mlir_tablegen_target(MLIRSCFPassIncGen)
 
 add_mlir_doc(Passes SCFPasses ./ -gen-pass-doc)
 
diff --git a/include/mlir/Dialect/SPIRV/IR/CMakeLists.txt b/include/mlir/Dialect/SPIRV/IR/CMakeLists.txt
index de3148d0a42d..0443e458d194 100644
--- a/include/mlir/Dialect/SPIRV/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/SPIRV/IR/CMakeLists.txt
@@ -4,35 +4,29 @@ add_mlir_doc(SPIRVOps SPIRVOps Dialects/ -gen-op-doc)
 set(LLVM_TARGET_DEFINITIONS SPIRVBase.td)
 mlir_tablegen(SPIRVEnums.h.inc -gen-enum-decls)
 mlir_tablegen(SPIRVEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRSPIRVEnumsIncGen)
-add_dependencies(mlir-headers MLIRSPIRVEnumsIncGen)
+add_mlir_tablegen_target(MLIRSPIRVEnumsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS SPIRVBase.td)
 mlir_tablegen(SPIRVEnumAvailability.h.inc -gen-spirv-enum-avail-decls)
 mlir_tablegen(SPIRVEnumAvailability.cpp.inc -gen-spirv-enum-avail-defs)
 mlir_tablegen(SPIRVCapabilityImplication.inc -gen-spirv-capability-implication)
-add_public_tablegen_target(MLIRSPIRVEnumAvailabilityIncGen)
-add_dependencies(mlir-headers MLIRSPIRVEnumAvailabilityIncGen)
+add_mlir_tablegen_target(MLIRSPIRVEnumAvailabilityIncGen)
 
 set(LLVM_TARGET_DEFINITIONS SPIRVOps.td)
 mlir_tablegen(SPIRVAvailability.h.inc -gen-avail-interface-decls)
 mlir_tablegen(SPIRVAvailability.cpp.inc -gen-avail-interface-defs)
 mlir_tablegen(SPIRVOpAvailabilityImpl.inc -gen-spirv-avail-impls)
-add_public_tablegen_target(MLIRSPIRVAvailabilityIncGen)
-add_dependencies(mlir-headers MLIRSPIRVAvailabilityIncGen)
+add_mlir_tablegen_target(MLIRSPIRVAvailabilityIncGen)
 
 set(LLVM_TARGET_DEFINITIONS SPIRVOps.td)
 mlir_tablegen(SPIRVSerialization.inc -gen-spirv-serialization)
-add_public_tablegen_target(MLIRSPIRVSerializationGen)
-add_dependencies(mlir-headers MLIRSPIRVSerializationGen)
+add_mlir_tablegen_target(MLIRSPIRVSerializationGen)
 
 set(LLVM_TARGET_DEFINITIONS SPIRVBase.td)
 mlir_tablegen(SPIRVAttrUtils.inc -gen-spirv-attr-utils)
-add_public_tablegen_target(MLIRSPIRVAttrUtilsGen)
-add_dependencies(mlir-headers MLIRSPIRVAttrUtilsGen)
+add_mlir_tablegen_target(MLIRSPIRVAttrUtilsGen)
 
 set(LLVM_TARGET_DEFINITIONS SPIRVAttributes.td)
 mlir_tablegen(SPIRVAttributes.h.inc -gen-attrdef-decls)
 mlir_tablegen(SPIRVAttributes.cpp.inc -gen-attrdef-defs)
-add_public_tablegen_target(MLIRSPIRVAttributeIncGen)
-add_dependencies(mlir-headers MLIRSPIRVAttributeIncGen)
+add_mlir_tablegen_target(MLIRSPIRVAttributeIncGen)
diff --git a/include/mlir/Dialect/Shape/Transforms/CMakeLists.txt b/include/mlir/Dialect/Shape/Transforms/CMakeLists.txt
index 83a500d7b0ad..db5539c2a10c 100644
--- a/include/mlir/Dialect/Shape/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Shape/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Shape)
-add_public_tablegen_target(MLIRShapeTransformsIncGen)
+add_mlir_tablegen_target(MLIRShapeTransformsIncGen)
 
 add_mlir_doc(Passes ShapePasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/SparseTensor/IR/CMakeLists.txt b/include/mlir/Dialect/SparseTensor/IR/CMakeLists.txt
index 54ad9491cce5..e583ccca4117 100644
--- a/include/mlir/Dialect/SparseTensor/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/SparseTensor/IR/CMakeLists.txt
@@ -6,15 +6,14 @@ mlir_tablegen(SparseTensorAttrDefs.h.inc -gen-attrdef-decls)
 mlir_tablegen(SparseTensorAttrDefs.cpp.inc -gen-attrdef-defs)
 mlir_tablegen(SparseTensorAttrEnums.h.inc -gen-enum-decls)
 mlir_tablegen(SparseTensorAttrEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRSparseTensorAttrDefsIncGen)
+add_mlir_tablegen_target(MLIRSparseTensorAttrDefsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS SparseTensorTypes.td)
 mlir_tablegen(SparseTensorTypes.h.inc -gen-typedef-decls)
 mlir_tablegen(SparseTensorTypes.cpp.inc -gen-typedef-defs)
-add_public_tablegen_target(MLIRSparseTensorTypesIncGen)
+add_mlir_tablegen_target(MLIRSparseTensorTypesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS SparseTensorInterfaces.td)
 mlir_tablegen(SparseTensorInterfaces.h.inc -gen-op-interface-decls)
 mlir_tablegen(SparseTensorInterfaces.cpp.inc -gen-op-interface-defs)
-add_public_tablegen_target(MLIRSparseTensorInterfacesIncGen)
-add_dependencies(mlir-headers MLIRSparseTensorInterfacesIncGen)
+add_mlir_tablegen_target(MLIRSparseTensorInterfacesIncGen)
diff --git a/include/mlir/Dialect/SparseTensor/TransformOps/CMakeLists.txt b/include/mlir/Dialect/SparseTensor/TransformOps/CMakeLists.txt
index dc29a0b41572..83751666f0f0 100644
--- a/include/mlir/Dialect/SparseTensor/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/SparseTensor/TransformOps/CMakeLists.txt
@@ -1,4 +1,4 @@
 set(LLVM_TARGET_DEFINITIONS SparseTensorTransformOps.td)
 mlir_tablegen(SparseTensorTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(SparseTensorTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRSparseTensorTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRSparseTensorTransformOpsIncGen)
diff --git a/include/mlir/Dialect/SparseTensor/Transforms/CMakeLists.txt b/include/mlir/Dialect/SparseTensor/Transforms/CMakeLists.txt
index a24a94705bc8..a3f89a56a981 100644
--- a/include/mlir/Dialect/SparseTensor/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/SparseTensor/Transforms/CMakeLists.txt
@@ -2,5 +2,5 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name SparseTensor)
 mlir_tablegen(Passes.capi.h.inc -gen-pass-capi-header --prefix SparseTensor)
 mlir_tablegen(Passes.capi.cpp.inc -gen-pass-capi-impl --prefix SparseTensor)
-add_public_tablegen_target(MLIRSparseTensorPassIncGen)
+add_mlir_tablegen_target(MLIRSparseTensorPassIncGen)
 add_mlir_doc(Passes SparseTensorPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Tensor/TransformOps/CMakeLists.txt b/include/mlir/Dialect/Tensor/TransformOps/CMakeLists.txt
index bb9f703602dc..735af942b540 100644
--- a/include/mlir/Dialect/Tensor/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/Tensor/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS TensorTransformOps.td)
 mlir_tablegen(TensorTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(TensorTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRTensorTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRTensorTransformOpsIncGen)
 
 add_mlir_doc(TensorTransformOps TensorTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Tensor/Transforms/CMakeLists.txt b/include/mlir/Dialect/Tensor/Transforms/CMakeLists.txt
index b312cee09930..dbe36037d46b 100644
--- a/include/mlir/Dialect/Tensor/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Tensor/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Tensor)
-add_public_tablegen_target(MLIRTensorTransformsIncGen)
+add_mlir_tablegen_target(MLIRTensorTransformsIncGen)
 
 add_mlir_doc(Passes TensorPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Tosa/Transforms/CMakeLists.txt b/include/mlir/Dialect/Tosa/Transforms/CMakeLists.txt
index d4e266183831..bbc8c75f5c47 100644
--- a/include/mlir/Dialect/Tosa/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Tosa/Transforms/CMakeLists.txt
@@ -2,7 +2,6 @@ set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name TosaOpt)
 mlir_tablegen(PassesEnums.h.inc -gen-enum-decls)
 mlir_tablegen(PassesEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRTosaPassIncGen)
-add_dependencies(mlir-headers MLIRTosaPassIncGen)
+add_mlir_tablegen_target(MLIRTosaPassIncGen)
 
 add_mlir_doc(Passes TosaPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/Transform/DebugExtension/CMakeLists.txt b/include/mlir/Dialect/Transform/DebugExtension/CMakeLists.txt
index 2b49faef7f00..eb84a627eaac 100644
--- a/include/mlir/Dialect/Transform/DebugExtension/CMakeLists.txt
+++ b/include/mlir/Dialect/Transform/DebugExtension/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS DebugExtensionOps.td)
 mlir_tablegen(DebugExtensionOps.h.inc -gen-op-decls)
 mlir_tablegen(DebugExtensionOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRTransformDialectDebugExtensionOpsIncGen)
+add_mlir_tablegen_target(MLIRTransformDialectDebugExtensionOpsIncGen)
 
 add_mlir_doc(DebugExtensionOps DebugExtensionOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Transform/IRDLExtension/CMakeLists.txt b/include/mlir/Dialect/Transform/IRDLExtension/CMakeLists.txt
index dfcd906b43af..44491a4ff9ae 100644
--- a/include/mlir/Dialect/Transform/IRDLExtension/CMakeLists.txt
+++ b/include/mlir/Dialect/Transform/IRDLExtension/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS IRDLExtensionOps.td)
 mlir_tablegen(IRDLExtensionOps.h.inc -gen-op-decls)
 mlir_tablegen(IRDLExtensionOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRTransformDialectIRDLExtensionOpsIncGen)
+add_mlir_tablegen_target(MLIRTransformDialectIRDLExtensionOpsIncGen)
 
 add_mlir_doc(IRDLExtensionOps IRDLExtensionOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Transform/Interfaces/CMakeLists.txt b/include/mlir/Dialect/Transform/Interfaces/CMakeLists.txt
index 14ce5b82b811..613d5c312866 100644
--- a/include/mlir/Dialect/Transform/Interfaces/CMakeLists.txt
+++ b/include/mlir/Dialect/Transform/Interfaces/CMakeLists.txt
@@ -6,8 +6,7 @@ add_mlir_doc(TransformInterfaces TransformOpInterfaces Dialects/ -gen-op-interfa
 set(LLVM_TARGET_DEFINITIONS TransformInterfaces.td)
 mlir_tablegen(TransformTypeInterfaces.h.inc -gen-type-interface-decls)
 mlir_tablegen(TransformTypeInterfaces.cpp.inc -gen-type-interface-defs)
-add_public_tablegen_target(MLIRTransformDialectTypeInterfacesIncGen)
-add_dependencies(mlir-headers MLIRTransformDialectTypeInterfacesIncGen)
+add_mlir_tablegen_target(MLIRTransformDialectTypeInterfacesIncGen)
 add_mlir_doc(TransformInterfaces TransformTypeInterfaces Dialects/ -gen-type-interface-docs)
 
 add_mlir_interface(MatchInterfaces)
diff --git a/include/mlir/Dialect/Transform/LoopExtension/CMakeLists.txt b/include/mlir/Dialect/Transform/LoopExtension/CMakeLists.txt
index 8f5e510ad39a..64e8b24872bb 100644
--- a/include/mlir/Dialect/Transform/LoopExtension/CMakeLists.txt
+++ b/include/mlir/Dialect/Transform/LoopExtension/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS LoopExtensionOps.td)
 mlir_tablegen(LoopExtensionOps.h.inc -gen-op-decls)
 mlir_tablegen(LoopExtensionOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRTransformDialectLoopExtensionOpsIncGen)
+add_mlir_tablegen_target(MLIRTransformDialectLoopExtensionOpsIncGen)
 
 add_mlir_doc(LoopExtensionOps LoopExtensionOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Transform/PDLExtension/CMakeLists.txt b/include/mlir/Dialect/Transform/PDLExtension/CMakeLists.txt
index 6af6b838f266..d887ed14e5cb 100644
--- a/include/mlir/Dialect/Transform/PDLExtension/CMakeLists.txt
+++ b/include/mlir/Dialect/Transform/PDLExtension/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS PDLExtensionOps.td)
 mlir_tablegen(PDLExtensionOps.h.inc -gen-op-decls)
 mlir_tablegen(PDLExtensionOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRTransformDialectPDLExtensionOpsIncGen)
+add_mlir_tablegen_target(MLIRTransformDialectPDLExtensionOpsIncGen)
 
 add_mlir_doc(PDLExtensionOps PDLExtensionOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Transform/Transforms/CMakeLists.txt b/include/mlir/Dialect/Transform/Transforms/CMakeLists.txt
index 3a399e6e0513..e931ebae5279 100644
--- a/include/mlir/Dialect/Transform/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Transform/Transforms/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Transform)
-add_public_tablegen_target(MLIRTransformDialectTransformsIncGen)
+add_mlir_tablegen_target(MLIRTransformDialectTransformsIncGen)
 
 add_mlir_doc(Passes TransformPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/UB/IR/CMakeLists.txt b/include/mlir/Dialect/UB/IR/CMakeLists.txt
index 0449cb2e9a9c..f02c15fa0e21 100644
--- a/include/mlir/Dialect/UB/IR/CMakeLists.txt
+++ b/include/mlir/Dialect/UB/IR/CMakeLists.txt
@@ -5,11 +5,11 @@ mlir_tablegen(UBOpsDialect.h.inc -gen-dialect-decls)
 mlir_tablegen(UBOpsDialect.cpp.inc -gen-dialect-defs)
 mlir_tablegen(UBOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=ub)
 mlir_tablegen(UBOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=ub)
-add_public_tablegen_target(MLIRUBOpsIncGen)
+add_mlir_tablegen_target(MLIRUBOpsIncGen)
 
 add_mlir_doc(UBOps UBOps Dialects/ -gen-dialect-doc)
 
 set(LLVM_TARGET_DEFINITIONS UBOpsInterfaces.td)
 mlir_tablegen(UBOpsInterfaces.h.inc -gen-attr-interface-decls)
 mlir_tablegen(UBOpsInterfaces.cpp.inc -gen-attr-interface-defs)
-add_public_tablegen_target(MLIRUBOpsInterfacesIncGen)
+add_mlir_tablegen_target(MLIRUBOpsInterfacesIncGen)
diff --git a/include/mlir/Dialect/Utils/CMakeLists.txt b/include/mlir/Dialect/Utils/CMakeLists.txt
index edfb1ca873ab..a60a0ba72f9a 100644
--- a/include/mlir/Dialect/Utils/CMakeLists.txt
+++ b/include/mlir/Dialect/Utils/CMakeLists.txt
@@ -1,5 +1,4 @@
 set(LLVM_TARGET_DEFINITIONS StructuredOpsUtils.td)
 mlir_tablegen(DialectUtilsEnums.h.inc -gen-enum-decls)
 mlir_tablegen(DialectUtilsEnums.cpp.inc -gen-enum-defs)
-add_public_tablegen_target(MLIRDialectUtilsIncGen)
-add_dependencies(mlir-headers MLIRDialectUtilsIncGen)
+add_mlir_tablegen_target(MLIRDialectUtilsIncGen)
diff --git a/include/mlir/Dialect/Vector/TransformOps/CMakeLists.txt b/include/mlir/Dialect/Vector/TransformOps/CMakeLists.txt
index ec8b0f4044d7..f8dd3f603961 100644
--- a/include/mlir/Dialect/Vector/TransformOps/CMakeLists.txt
+++ b/include/mlir/Dialect/Vector/TransformOps/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(LLVM_TARGET_DEFINITIONS VectorTransformOps.td)
 mlir_tablegen(VectorTransformOps.h.inc -gen-op-decls)
 mlir_tablegen(VectorTransformOps.cpp.inc -gen-op-defs)
-add_public_tablegen_target(MLIRVectorTransformOpsIncGen)
+add_mlir_tablegen_target(MLIRVectorTransformOpsIncGen)
 
 add_mlir_doc(VectorTransformOps VectorTransformOps Dialects/ -gen-op-doc)
diff --git a/include/mlir/Dialect/Vector/Transforms/CMakeLists.txt b/include/mlir/Dialect/Vector/Transforms/CMakeLists.txt
index 2c288fe3b77f..2671706d5258 100644
--- a/include/mlir/Dialect/Vector/Transforms/CMakeLists.txt
+++ b/include/mlir/Dialect/Vector/Transforms/CMakeLists.txt
@@ -4,6 +4,6 @@ mlir_tablegen(VectorTransformsEnums.cpp.inc -gen-enum-defs)
 
 set(LLVM_TARGET_DEFINITIONS Passes.td)
 mlir_tablegen(Passes.h.inc -gen-pass-decls -name Vector)
-add_public_tablegen_target(MLIRVectorTransformsIncGen)
+add_mlir_tablegen_target(MLIRVectorTransformsIncGen)
 
 add_mlir_doc(Passes VectorPasses ./ -gen-pass-doc)
diff --git a/include/mlir/Dialect/X86Vector/CMakeLists.txt b/include/mlir/Dialect/X86Vector/CMakeLists.txt
index a22f8332514b..32127fdc6ff2 100644
--- a/include/mlir/Dialect/X86Vector/CMakeLists.txt
+++ b/include/mlir/Dialect/X86Vector/CMakeLists.txt
@@ -3,4 +3,4 @@ add_mlir_doc(X86Vector X86Vector Dialects/ -gen-dialect-doc -dialect=x86vector)
 
 set(LLVM_TARGET_DEFINITIONS X86Vector.td)
 mlir_tablegen(X86VectorConversions.inc -gen-llvmir-conversions)
-add_public_tablegen_target(MLIRX86VectorConversionsIncGen)
+add_mlir_tablegen_target(MLIRX86VectorConversionsIncGen)
diff --git a/include/mlir/IR/CMakeLists.txt b/include/mlir/IR/CMakeLists.txt
index 0c7937dfd69e..54681413b5b9 100644
--- a/include/mlir/IR/CMakeLists.txt
+++ b/include/mlir/IR/CMakeLists.txt
@@ -13,48 +13,58 @@ set(LLVM_TARGET_DEFINITIONS BuiltinAttributes.td)
 mlir_tablegen(BuiltinAttributes.h.inc -gen-attrdef-decls)
 mlir_tablegen(BuiltinAttributes.cpp.inc -gen-attrdef-defs)
 add_public_tablegen_target(MLIRBuiltinAttributesIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinAttributesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinAttributeInterfaces.td)
 mlir_tablegen(BuiltinAttributeInterfaces.h.inc -gen-attr-interface-decls)
 mlir_tablegen(BuiltinAttributeInterfaces.cpp.inc -gen-attr-interface-defs)
 add_public_tablegen_target(MLIRBuiltinAttributeInterfacesIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinAttributeInterfacesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinDialect.td)
 mlir_tablegen(BuiltinDialect.h.inc -gen-dialect-decls)
 mlir_tablegen(BuiltinDialect.cpp.inc -gen-dialect-defs)
 add_public_tablegen_target(MLIRBuiltinDialectIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinDialectIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinDialectBytecode.td)
 mlir_tablegen(BuiltinDialectBytecode.cpp.inc -gen-bytecode -bytecode-dialect="Builtin")
 add_public_tablegen_target(MLIRBuiltinDialectBytecodeIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinDialectBytecodeIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinLocationAttributes.td)
 mlir_tablegen(BuiltinLocationAttributes.h.inc -gen-attrdef-decls)
 mlir_tablegen(BuiltinLocationAttributes.cpp.inc -gen-attrdef-defs)
 add_public_tablegen_target(MLIRBuiltinLocationAttributesIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinLocationAttributesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinOps.td)
 mlir_tablegen(BuiltinOps.h.inc -gen-op-decls)
 mlir_tablegen(BuiltinOps.cpp.inc -gen-op-defs)
 add_public_tablegen_target(MLIRBuiltinOpsIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinOpsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinTypes.td)
 mlir_tablegen(BuiltinTypes.h.inc -gen-typedef-decls)
 mlir_tablegen(BuiltinTypes.cpp.inc -gen-typedef-defs)
 add_public_tablegen_target(MLIRBuiltinTypesIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinTypesIncGen)
 mlir_tablegen(BuiltinTypeConstraints.h.inc -gen-type-constraint-decls)
 mlir_tablegen(BuiltinTypeConstraints.cpp.inc -gen-type-constraint-defs)
 add_public_tablegen_target(MLIRBuiltinTypeConstraintsIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinTypeConstraintsIncGen)
 
 set(LLVM_TARGET_DEFINITIONS BuiltinTypeInterfaces.td)
 mlir_tablegen(BuiltinTypeInterfaces.h.inc -gen-type-interface-decls)
 mlir_tablegen(BuiltinTypeInterfaces.cpp.inc -gen-type-interface-defs)
 add_public_tablegen_target(MLIRBuiltinTypeInterfacesIncGen)
+add_dependencies(mlir-generic-headers MLIRBuiltinTypeInterfacesIncGen)
 
 set(LLVM_TARGET_DEFINITIONS TensorEncoding.td)
 mlir_tablegen(TensorEncInterfaces.h.inc -gen-attr-interface-decls)
 mlir_tablegen(TensorEncInterfaces.cpp.inc -gen-attr-interface-defs)
 add_public_tablegen_target(MLIRTensorEncodingIncGen)
+add_dependencies(mlir-generic-headers MLIRTensorEncodingIncGen)
 
 add_mlir_doc(BuiltinAttributes BuiltinAttributes Dialects/ -gen-attrdef-doc)
 add_mlir_doc(BuiltinLocationAttributes BuiltinLocationAttributes Dialects/ -gen-attrdef-doc)
diff --git a/tools/mlir-shlib/CMakeLists.txt b/tools/mlir-shlib/CMakeLists.txt
index a33c70c5807b..166bab744de3 100644
--- a/tools/mlir-shlib/CMakeLists.txt
+++ b/tools/mlir-shlib/CMakeLists.txt
@@ -30,7 +30,7 @@ if(MLIR_LINK_MLIR_DYLIB)
   set(INSTALL_WITH_TOOLCHAIN INSTALL_WITH_TOOLCHAIN)
 endif()
 
-if(LLVM_BUILD_LLVM_DYLIB)
+if(MLIR_BUILD_MLIR_DYLIB)
   add_mlir_library(
     MLIR
     SHARED
@@ -45,6 +45,7 @@ if(LLVM_BUILD_LLVM_DYLIB)
     ${mlir_llvm_link_components}
   )
   target_link_libraries(MLIR PRIVATE ${LLVM_PTHREAD_LIB})
+  target_link_libraries(MLIRDylib INTERFACE MLIR)
 endif()
 
 #message("Libraries included in libMLIR.so: ${mlir_libs}")
